### PR TITLE
Stacked/unified block toolbar

### LINKED ISSUE #26114: Keyboard navigation to image caption toolbar
**Describe the bug**
Cannot navigate into the image caption toolbar with the keyboard.

**To reproduce**
1. Create image block
2. Enter some text
3. Try to change that text color with just the keyboard
4. Unable to navigate into toolbar

**Expected behavior**
Should be able to navigate into the image caption toolbar with the keyboard.

**Screenshots**
caption toolbar

**Editor version (please complete the following information):**
- WordPress version: 5.5.1
- Does the website has Gutenberg plugin installed, or is it using the block editor that comes by default? Gutenberg plugin
- If the Gutenberg plugin is installed, which version is it? Master branch (b126b3f)

**Desktop:**
 - OS: MacOS
 - Browser: Chrome
 - Version: 85.0.4183.121

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Best reviewed [without whitespace changes](https://github.com/WordPress/gutenberg/pull/31134/files?diff=split&w=1).

Alternative to #30234.
Also fixes #26114, a long standing accessibility issue with the inline rich text toolbar.

Also unifies toolbar APIs under one simple `<BlockTools />` component and removes the need for implementors to render a dedicated block toolbar popover slot.

unified-toolbar

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/README.md [hunk 1]
```diff
@@ -34,7 +35,7 @@ function MyEditorComponent() {
 			onChange={ ( blocks ) => updateBlocks( blocks ) }
 		>
 			<SlotFillProvider>
-				<Popover.Slot name="block-toolbar" />
```
Comment (ellatrix) [inline packages/block-editor/README.md:37]: @youknowriad I think you'll be happy to see this 😄

File: packages/block-editor/src/components/block-list/block-contextual-toolbar.js [hunk 2]
```diff
@@ -41,22 +46,24 @@ function BlockContextualToolbar( { focusOnMount, ...props } ) {
...
+	const classes = classnames(
+		'block-editor-block-contextual-toolbar',
+		className,
+		{
+			'has-parent': hasParents,
+		}
+	);
 
 	return (
-		<div className="block-editor-block-contextual-toolbar-wrapper">
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/block-contextual-toolbar.js]: This class is not used anywhere, so it should be safe to remove this useless div.

File: packages/block-editor/src/components/block-toolbar/unified.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+export default function UnifiedBlockToolbar() {
+	const hasFixedToolbar = useSelect(
+		( select ) => select( blockEditorStore ).getSettings().hasFixedToolbar,
+		[]
+	);
+
+	if ( hasFixedToolbar ) {
+		return (
+			<>
+				<Popover.Slot name="block-toolbar" />
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-toolbar/unified.js]: Do we still need this slot? I have trouble understanding why this `UnifiedBlockToolbar` doesn't directly render `BlockContextualToolbar` in all  cases?

File: packages/block-editor/src/components/block-list/style.scss [hunk 4]
```diff
@@ -512,6 +512,25 @@
 		border-right-color: $gray-900;
 	}
 
+	&.is-fixed {
+		position: sticky;
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/style.scss]: I feel the is-fixed styles and logic are misplaced, they make more sense  in the `edit-*` packages because we don't really know the container here and these styles depend on their container right?

File: packages/block-editor/src/components/block-toolbar/unified.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-toolbar/unified.js]: Because the block popover needs access to the block nodes, which is only available in the context of the BlockList. I plan to change this in the future though, so then we'll be able to render it here in place.

Comment (ellatrix) [reply packages/block-editor/src/components/block-toolbar/unified.js]: The important thing is that the slot becomes just an internal thing.

File: packages/block-editor/src/components/block-list/style.scss (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/style.scss]: Sure, we could move it. I don't mind either way. `sticky` just falls back to being relative if there's no scroll container though.

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/style.scss]: For example, in the site editor there's no scroll container, but the content box below is scrollable.

**Comment by draganescu** [comment]
I love this so much! 👏🏻

**Comment by jasmussen** [comment]
Thanks so much for working on this, it's a big one! Mostly I love the huge amount of red code in this one. Let me dive in.

Here's the post editor before:

post editor before

Here's after:

post editor after

Here's the site editor before:

site editor before

Here's the site editor after:

site editor after

I'm seeing one issue:

bottom padding

Note how there's always a little extra scrollspace at the bottom. I'm also seeing this in the site editor, but it's less pronounced there. Let me know if you need help debugging this.

---

Let's get back to the "Why" of this PR, especially if you're used to seeing this:

Screenshot 2021-04-26 at 11 04 59

Because it isn't immediately obvious why this is better:

Screenshot 2021-04-26 at 11 06 02

The empty toolbar also takes up space even when you have unselected blocks:

Screenshot 2021-04-26 at 11 06 37

Those are good questions, and they are opportunities to improve the overall design of the top toolbar. There's in fact a dedicated issue for it, #20592, wherein I think we should explore those solutions. 

However here's why I think it's still important to land this PR today. Firstly, the toolbar is entirely broken in the site editor:

b

Even when the toolbar arbitrarily stacks, it covers content:

Screenshot 2021-04-26 at 11 13 56

The tabbing order is also broken as soon as stacked:

a

This PR fixes all that:

- It doesn't cover content
- The toolbar is consistently placed, stacked always from mobile to desktop
- The toolbar does not get cropped
- Tabbing order matches visual order
- The implementation seems vastly simpler

For that reason, I'm a fan, and I think we can follow up on improving the visuals even further as part of #20592.

**Comment by ellatrix** [comment]
It makes sense both visually, for accessibility and in the code. I'll look into the issues in a bit.

**Comment by jasmussen** [comment]
One other thing, in the post editor, the scrollbar starts at the top:

Screenshot 2021-04-26 at 14 12 07

In the site editor, it starts under the white bar. The latter is ideal, the former less so. But I presume that if/when the iframe could land for the post editor, we'd get the latter there as well?

**Comment by ellatrix** [comment]
@jasmussen What do you mean, do you not expect it to start at the top? There's no way to avoid that with sticky positioning. That's just how sticky works.

File: packages/block-editor/src/components/block-toolbar/unified.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+				<BlockContextualToolbar hideDragHandle className="is-fixed" />
```
Comment (talldan) [inline packages/block-editor/src/components/block-toolbar/unified.js]: Minor but I think it'd be preferable to make this an `isFixed` prop. 

Potentially, `hideDragHandle` won't be needed then, as I imagine it's only true when `isFixed` is also true.

File: packages/block-editor/src/components/block-toolbar/unified.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import { store as blockEditorStore } from '../../store';
+import BlockContextualToolbar from '../block-list/block-contextual-toolbar';
+
+export default function UnifiedBlockToolbar() {
+	const hasFixedToolbar = useSelect(
+		( select ) => select( blockEditorStore ).getSettings().hasFixedToolbar,
+		[]
+	);
+
+	if ( hasFixedToolbar ) {
```
Comment (talldan) [inline packages/block-editor/src/components/block-toolbar/unified.js]: Testing this PR and on mobile viewports (with Top Toolbar disabled), there's no block toolbar. 

My understanding is that the desired behavior is for the fixed toolbar to show on mobile viewports.

I think there are still a few files that have some viewport logic for hiding the floating toolbar on mobile viewports (BlockPopover, BlockToolbar).

I wonder if as much of that viewport logic could be consolidated into one place, possibly this component and this `if` statement.

One of the difficult parts of implementing an editor is that the mobile version of the block toolbar doesn't work out of the box, a developer has to specifically add the fixed toolbar even if they're not implementing top toolbar mode, so it'd be nice if `UnifiedBlockToolbar` helped solve that.

At the same time, if an implementor wants the responsiveness of the toolbar to behave differently to the post editor, that's difficult to do as the viewport logic is implicit to much of the block editor package. I think ideally `BlockToolbar` would be just a plain configurable toolbar that we build abstractions around.

**Review by talldan — COMMENTED** [review]
I'm definitely in favour of this, especially the removal of all those styles.

**Comment by jasmussen** [comment]
This is looking great:

post editor

site editor

responsive

There's still a small thing with the scrollbar in the site editor, I'm taking a look now. 

Dan caught an important thing to fix above, here's a GIF showing it:

mobile

I.e. the toolbar should always be docked at the top on the mobile breakpoint.

**Comment by jasmussen** [comment]
I debugged a bit and it appears that the extra scrollheight is present in the site editor in trunk:

trunk

☝️ check out how bad the overlap situation is in the site editor currently 🙈 

And is fixed by applying `box-sizing: border-box;` to the `body` element inside the iframe. 

Worth fixing, right? But doesn't necessarily have to be this PR, so it seems like Dan's issue is the one to fix and then we can land this.

File: packages/block-editor/src/components/block-toolbar/unified.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-toolbar/unified.js]: Once we have #29573, this will be possible.

File: packages/block-editor/src/components/block-toolbar/unified.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-toolbar/unified.js]: Sure, sounds good

**Comment by ellatrix** [comment]
@jasmussen I fixed the viewport issue. Not sure how to reproduce the other issue. It's strange that we need to change box sizing because anything in the iframe shouldn't affect the outside. Let's investigate this in a separate PR if it's happening in trunk too?

**Comment by ellatrix** [comment]
@jasmussen Btw, looking at your GIF, I see a rule `100% + 32px` being applied for full screen mode. Could that have something to do with it?

**Comment by jasmussen** [comment]
It does happen in trunk yes.

> Btw, looking at your GIF, I see a rule 100% + 32px being applied for full screen mode. Could that have something to do with it?

Yes that could very well have something to do with it. It reminds me of some of the adminbar CSS that WordPress applies. It would most definitely be that if it was 100% + 46px at the less than <600px breakpoint.

File: packages/block-editor/src/components/block-toolbar/unified.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-toolbar/unified.js]: Rebase and moved the insertion point and popover toolbar into the unified component.

**Review by tellthemachines — APPROVED** [review]
Tested after the recent changes and it's working very well! Huge improvement to keyboard navigation both with top toolbar enabled and for the inline formatting toolbar 🎉

**Comment by talldan** [comment]
There do seem to be some changes in other editors. Not sure I'd expect this PR to fix everything given its already quite big, but just thought I'd make a note of what I'm seeing:

### Customize Widgets editor (for testing I cherry picked the fix from https://github.com/WordPress/gutenberg/pull/31262, which was unmerged at the time):
- An error is being thrown in the console (`Uncaught TypeError: Cannot read property 'current' of undefined
    at HTMLDivElement.onMouseMove (use-in-between-inserter.js:32)`). I think this might be because `BlockTools` is missing in this editor. I think block editors should still work without that component, otherwise we'll introduce problems for users of the package.
- The editor always now has a fixed unstyled toolbar. I think this is expected from looking at the code as the `BlockToolbar` is the only component used. There's already an issue for implementing the toolbar responsively (https://github.com/WordPress/gutenberg/issues/30720), so I think there will already be a follow up to this PR implementing `BlockTools` in the customize widgets editor.

Before:
Screenshot 2021-04-28 at 12 01 05 pm

After:
Screenshot 2021-04-28 at 12 01 39 pm

### Standalone widget editor
The 'fixed' toolbar here has changed position and will require a follow-up to fix some styles.

Before:
Screenshot 2021-04-28 at 11 55 48 am

After:
Screenshot 2021-04-28 at 11 57 01 am

### Navigation editor
The 'fixed' toolbar has changed position. It's now more consistent with the other editors, which might turn out well given https://github.com/WordPress/gutenberg/issues/31241 proposes making the editor toolbars consistent. Probably needs a bit of CSS polish (and there's a good chance there'll be CSS to remove). (cc @javierarce)

Before:
Screenshot 2021-04-28 at 11 58 23 am

After:
Screenshot 2021-04-28 at 11 57 58 am

**Comment by ellatrix** [comment]
I'll continue to work on this today. Make some changes yesterday that require some polish

**Comment by jasmussen** [comment]
@ellatrix — doesn't have to be part of this PR, but per [Matías thoughts in this comment](https://github.com/WordPress/gutenberg/pull/30983#issuecomment-827604003), with the direction this PR is going, can we do this:

Screenshot-2021-04-28-at-09 14 45

... instead of this:

Screenshot 2021-04-28 at 09 20 20

Specifically, carve out some padding at the top of the page with the same color as the theme, and then show the regular block toolbar just pinned in that top left area?

**Comment by ellatrix** [comment]
@jasmussen That was my initial thought, but how do you get to the content under the toolbar if the content isn't shifted down?

**Comment by jasmussen** [comment]
Just tried inspecting, it seems like that should be possible:

Screenshot 2021-04-28 at 09 26 32

In this one, I'm just abs positioning the toolbar, and adding top-padding to the `html` element inside the iframe. That last bit is probably the only contentious one, but I figere WordPress does the same for the wp-adminbar, so it might work?

**Comment by ellatrix** [comment]
@jasmussen Ok. This is just to have the background around the toolbar right? I'll try to make it work. It was my original thought to use the block popover, but then ran into issues to make the popover fill the entire horizontal width 😅 If this is what we want, we can reuse the popover again for the fixed toolbar.

**Comment by jasmussen** [comment]
Ack I'm sorry for the headaches :D 

I think it's important that in the top toolbar mode that we don't cover any content, that's one of the aspects that is super broken right now. But to still feel like the block toolbar, it'd be nice to have the theme background behind it. 

It sounds like a bigger shift in approach? Or an it land in a followup? Are there pros and cons to each approach? Which one do you like the best?

**Comment by jasmussen** [comment]
Note that detaching the toolbar as discussed would help fix #20592 as options c, d or e.

**Comment by ellatrix** [comment]
> It sounds like a bigger shift in approach? Or an it land in a followup? Are there pros and cons to each approach? Which one do you like the best?

Not at all I think, it should be easy to do. The padding is the annoying part, but we can use `calc` in CSS I guess. But there's benefits as well, we can reuse more parts of the block popover.

**Comment by ellatrix** [comment]
Ok, the navigation editor and the widget editor I will leave as it is now. These are minor styling changes that can be done as follow-ups and there's nothing broken. 

I still need to add the block tools to the widget editor in the customiser, but I can't figure out how to get the editor to load, even in trunk.

**Comment by jasmussen** [comment]
> I still need to add the block tools to the widget editor in the customiser, but I can't figure out how to get the editor to load, even in trunk.

The customizer is hidden in the menu when you're using a block theme.

**Comment by ellatrix** [comment]
@jasmussen I found it, it was the disable CSS animations plugin hiding it

**Comment by jasmussen** [comment]
This PR is working really well with TT1 blocks:

tt1

I'm seeing an issue in my own theme, which is very likely my own fault, but I'm curious what's up as it affects the appender:

hey

The height of the appender is right because I've added a ton of margins between the blocks on that page, so it's supposed to be tall. But it's pushed upwards some amount.

Screenshot 2021-04-28 at 11 26 32

The reason I'm showing it is because I'm not seeing that issue at all in trunk:

trunk

I do recall recently an issue with the appenders getting offset upwards and causing such issues, which seems similar. Could it have reappeared in this branch?

**Comment by ellatrix** [comment]
I fixed the insertion point. The iframe wasn't offset because the insertion point now renders rightfully outside the block list, but since we pass `getAnchorRect` instead of `anchorRef`, it's missing `ownerDocument` context, so we have to pass it through `getAnchorRect`.

**Comment by ellatrix** [comment]
This is ready for more review. I think it's looking good.

@youknowriad One thing we have to decide is how to handle the change. Should we let `BlockList` render the block tools if it's not one of its parent components (detection through context)? We can log a warning to say that using `<Popover.Slot name="block-toolbar" />` is deprecated, but still render the toolbars in that slot. Or are we fine with a breaking change?

**Comment by ellatrix** [comment]
Ok, I added a backward compatibility layer so that if there's an editor that doesn't `<BlockTools>` but `<Popover.Slot name="block-toolbar" />`, things will keep working and a deprecation message will be logged. I tested this with a few editors by removing the new component and restoring the slot and it works great.

**Comment by tellthemachines** [comment]
Hmm, sticky positioning isn't working in customize-widgets on mobile:

Screen Shot 2021-04-29 at 10 28 19 am

Edit: this is likely because several of its parents have `overflow` set to a non-default value. Changing these values messes with the whole sidebar layout, so we may have to make it `position: fixed` instead.

**Comment by ellatrix** [comment]
@tellthemachines I have to check trunk, but the toolbar was previously fixed too, so I preserved the styling. The fixed toolbar is also not enabled there (no option and no small viewport), so I don't think sticky positioning is enabled at all.

**Comment by tellthemachines** [comment]
In trunk, the toolbar in the customizer isn't showing at all on mobile, so this PR is already an improvement on that 😄 
It'll need a few styling tweaks to get it to stick to the top of the viewport, but that can be fixed separately - we already have #30720 to track it.

File: packages/block-editor/src/components/block-tools/index.js [hunk 7]
```diff
@@ -1,14 +1,46 @@
...
 export default function BlockTools( { children } ) {
+	const isLargeViewport = useViewportMatch( 'medium' );
+	const hasFixedToolbar = useSelect(
+		( select ) => select( blockEditorStore ).getSettings().hasFixedToolbar,
+		[]
+	);
+
 	return (
 		<InsertionPoint>
+			{ ( hasFixedToolbar || ! isLargeViewport ) && (
```
Comment (talldan) [inline packages/block-editor/src/components/block-tools/index.js:33]: In an earlier iteration this component was just implementing the two forms of the block toolbar, and I think it was good to include the viewport logic in that iteration. The `UnifiedBlockToolbar` component became a convenient and quick way to implement that kind of responsive toolbar.

But now this is bundled with some pretty key functionality for the block editor it seems fairly opinionated to say that someone implementing an editor should have a fixed toolbar on small viewports.

I don't think it's a blocker, this PR still makes things much easier overall. And an implementer can still reimplement this component without the viewport code. But I think it does pose a question about whether `BlockEditor` should be exporting high-level or low-level components.

Comment (ellatrix) [reply packages/block-editor/src/components/block-tools/index.js:33]: We can have a prop on this component `hasFixedToolbar` that overrides the setting? That's easy to add. :)

File: packages/edit-post/src/components/visual-editor/style.scss [hunk 8]
```diff
@@ -1,6 +1,7 @@
 .edit-post-visual-editor {
 	position: relative;
-
+	display: flex;
+	flex-flow: column;
```
Comment (talldan) [inline packages/edit-post/src/components/visual-editor/style.scss:4]: It doesn't seem to cause any issues, but I couldn't quite see why this change was made.

File: packages/block-editor/src/components/block-tools/index.js (hunk 7 above)
Comment (talldan) [reply packages/block-editor/src/components/block-tools/index.js:33]: No rush to do that right now, just some food for thought. 😄

**Review by talldan — APPROVED** [review]
This is working really well, great job @ellatrix 🎉 

I tested back compat by reverting the navigation editor's files back to `trunk` and it works perfectly.

The only issue I spotted is this behaviour when there are notices with the Top Toolbar or Fixed Toolbar, which would be good to tackle in a follow up:
scrolly-toolbar

I'll go ahead and merge and follow up myself with any needed adjustments for the widgets screens.

**Comment by jasmussen** [comment]
💯

File: packages/edit-post/src/components/visual-editor/style.scss (hunk 8 above)
Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/style.scss:4]: `flex-flow: column;` is a magical way to stack elements with the total height remaining 100%. If one of the elements sets its height to 100%, it will just fill up the remaining height. Cc @jasmussen

**Comment by ellatrix** [comment]
Thanks for the reviews!

File: packages/edit-post/src/components/visual-editor/style.scss (hunk 8 above)
Comment (jasmussen) [reply packages/edit-post/src/components/visual-editor/style.scss:4]: Very cool indeed. 

I'll keep an eye on any flex related issues that might occur from this, but in my testing this worked as intended, and it lets us avoid doing something like `height: calc(100% - #{ $block-toolbar-height } )` for the canvas.
